pkg/controller/registrysyncer/registrysyncer.go

```diff
@@ -258,44 +259,70 @@ func (r *reconciler) reconcile(ctx context.Context, req reconcile.Request, log *
 	return nil
 }
 
-func finalizeIfNeeded(ctx context.Context, stream *imagev1.ImageStream, srcClusterName string, clients map[string]ctrlruntimeclient.Client) error {
-	finalizerSet := sets.NewString(stream.Finalizers...)
-	if !finalizerSet.Has(finalizerName) {
-		return nil
+func finalizeIfNeeded(ctx context.Context, sourceImageStream *imagev1.ImageStream, srcClusterName string, clients map[string]ctrlruntimeclient.Client, log *logrus.Entry) (bool, error) {
+	deleteTimeStampClusters := sets.NewString()
```

Comment:
This seems overly complicated. Please just:
* Get the other imagestream
* Return if both have a nil DeletionTimestamp, without putting them in a different data structure first
* Iterate over them and
    * Remove finalizer if present
     * Delete

Reply:
This roughly does that. I have tried to simplify this but failed.
I guess this makes the function complicated:

1. handle differently with the one with DeleteTimestamp and without. For the former, we just remove the finalizer. For the latter, we remove the finalizer and delete. Maybe doing delete too for the former will make the function easy but probably we will get NotFound. Our concern yesterday was it was not fast enough. Maybe do only what we have to do. WDYT?
2. We have not assumed that we have only 2 clusters to sync. I agree that maybe we never need sync more than 2 clusters. But I do not want to make this function special if that is OK.

Reply:
> handle differently with the one with DeleteTimestamp and without. For the former, we just remove the finalizer. For the latter, we remove the finalizer and delete. Maybe doing delete too for the former will make the function easy but probably we will get NotFound. Our concern yesterday was it was not fast enough. Maybe do only what we have to do. WDYT?

That still doesn't require all the additional data structures:
```
stream := &imagev1.ImageStream{}
if err := client.Get(ctx, ctrlruntimeclient.ObjectKey{Name: sourceImageStream.Name, Namespace: sourceImageStream.Namespace}, stream); err != nil {
  if apierrors.IsNotFound{
     stream = nil
  }
  return false, err
}

if stream != nil && stream.DeletionTimestamp == nil && sourceImageStream.DeletionTimeStamp == nil {
  return false, nil
}
// finalize if needed stream
// finalize if needed sourceStream
// Delete if deletionTimestamp is unset for Stream, tolerate IsNotFound
// Delete if deletinTimestamp is unset for SourceStream, tolreate IsNotFound
```

> We have not assumed that we have only 2 clusters to sync. I agree that maybe we never need sync more than 2 clusters. But I do not want to make this function special if that is OK.

That is premature optimization, do not solve problems you don't have, especially not if they complicate your solution.

Reply:
Updated.